File: 0198-house-robber/0198-house-robber.cpp
class Solution {
public:
    int foo(vector<int>& nums,int index, vector<int>& dp){
        // DP Memo
        
        if(index>=nums.size()){
            return 0;
        }
        
        if(index== nums.size()-1){
            dp[index]=nums[index];
            return dp[index];
        }
        
        
        
        if(dp[index]!=-1){
            return dp[index];
        }
        
        dp[index]=max(nums[index]+foo(nums,index+2,dp),foo(nums,index+1,dp));
        return dp[index];

        // **** Recursive Solution. **** //
//         if(index>=nums.size()){
//             return 0;
//         }

//         if(index==nums.size()-1){
//             return nums[index];
//         }

//         if(nums.size()<=2){
//             return *max_element(nums.begin(),nums.end());
//         }

//         return max(nums[index]+foo(nums,index+2),foo(nums,index+1));
        
        
        
    }



    int rob(vector<int>& nums) {
        vector<int> dp(nums.size(),-1);
        int discard=foo(nums,0,dp);
        cout<<"discard val"<<discard<<endl;
        return dp[0];
    }
};

File: 0198-house-robber/0198-house-robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        n=len(nums)
        def foo(idx):
            if(idx>=n):
                return 0
            if(idx==n-1):
                return nums[idx]
            return max(nums[idx]+foo(idx+2), foo(idx+1))
        
        
        def foo2(nums):
            if(len(nums)<3):
                return max(nums)
            
            dp=[0]*(n+1)
            dp[0]=nums[0]
            dp[1]=max(nums[0],nums[1])
            
            for idx in range(2,n):
                dp[idx]=max(nums[idx]+dp[idx-2],dp[idx-1])
            
            return dp[n-1]
                
        return foo2(nums)